include/lbann/layers/transform/slice.hpp

```diff
@@ -236,256 +198,281 @@ class slice_layer : public transform_layer {
-          child->get_gpu_bp_output(input, work_spaces, this);
-        }
-        else {
-          child->get_bp_output(*this->m_prev_error_signal_v, this);
-          this->m_cudnn->scatter_to_gpus(work_spaces,
-                                         this->m_prev_error_signal_v->LockedMatrix(),
-                                         this->m_mini_batch_size_per_gpu);
-          input = work_spaces;
-        }
-      }
-
-      // Split previous error signal tensor into slices
-      const int input_slice_dim = m_slice_points[child_index+1] - m_slice_points[child_index];
-      const int input_slice_size = input_slice_dim * slice_unit_size;
-      const int input_size = num_slices * input_slice_size;
-      const int slice_offset = m_slice_points[child_index] * slice_unit_size;
-
-      // Copy slices from previous error signal tensor into error signal tensor
-      for(int slice = 0; slice < num_slices; ++slice) {
-        std::vector<DataType*> input_slice(num_gpus), output_slice(num_gpus);
-        for(int i = 0; i < num_gpus; ++i) {
-          input_slice[i] = input[i] + slice * input_slice_size;
-          output_slice[i] = this->m_error_signal_d[i] + slice * output_slice_size + slice_offset;
+  void fp_compute_cpu() {
+
+    // Input tensor
+    const auto& input = get_prev_activations();
+    const int width = input.Width();
+
+    // Get number of contiguous regions in a tensor slice of width 1
+    const auto& input_dims = this->m_prev_neuron_dims;
+    const int num_regions = std::accumulate(input_dims.begin(),
+                                            input_dims.begin() + m_slice_axis,
+                                            1,
+                                            std::multiplies<int>());
+    const int unit_region_size = std::accumulate(input_dims.begin() + m_slice_axis + 1,
+                                                 input_dims.end(),
+                                                 1,
+                                                 std::multiplies<int>());
+
+    // Get stride between contiguous regions in input tensor slices
+    const int input_slice_dim = input_dims[m_slice_axis];
+    const int input_region_stride = input_slice_dim * unit_region_size;
+    
+    // Populate output tensors with slices of input tensor
+    for (int i = 0; i < get_num_children(); ++i) {
+      auto& output = get_activations(i);
+
+      // Get stride between contiguous regions in output tensor slices
+      const int output_slice_dim = m_slice_points[i+1] - m_slice_points[i];
+      const int output_region_stride = output_slice_dim * unit_region_size;
+
+      // Get position of first contiguous region in input tensor
+      const int input_region_start = m_slice_points[i] * unit_region_size;
+      const int input_region_end = m_slice_points[i+1] * unit_region_size;
+
+      // Populate current output tensor
+      output.Resize(num_regions * output_region_stride, width);
+      if (num_regions == 1) {
```

Comment:
I believe that the num_regions trick being used here is the same thing as how we were using templated helper functions.  I still think that the num_regions is hard to remember when and why it works.

Reply:
This is needed in case the slice dimension is not the first tensor dimension, e.g. we slice an image along its rows instead of along its channels.